src/include/frag.h

```diff
@@ -567,29 +567,27 @@ class fragset_t {
       f = f.parent();
     }
   }
-  
+
+  void insert_raw(frag_t f){
+    _set.insert(f);
+  }
   void insert(frag_t f) {
     _set.insert(f);
     simplify();
   }
 
   void simplify() {
-    while (1) {
-      bool clean = true;
-      std::set<frag_t>::iterator p = _set.begin();
-      while (p != _set.end()) {
-	if (!p->is_root() &&
-	    _set.count(p->get_sibling())) {
-	  _set.erase(p->get_sibling());
-	  _set.insert(p->parent());
-	  _set.erase(p++);
-	  clean = false;
-	} else {
-	  p++;
-	}
+    std::set<frag_t>::iterator p = _set.begin();
+    while (p != _set.end()) {
+      if (!p->is_root() &&
+	  _set.count(p->get_sibling())) {
+	_set.erase(p->get_sibling());
+	_set.insert(p->parent());
+	_set.erase(p++);
+	--p;
```

Comment:
`p = _set.erase(p)` is simpler.

Reply:
auto ret = _set.insert(p->parent());
_set.erase(p)
p = ret.first;

Reply:
I assume you're relying on the ordered nature of the `std::set`? Please add a comment for this.

Reply:
maybe we can sort items in fragset in [frag_vale, frag_bit] order.  position of parent frag in set is close to its children. this will make the sort more efficient. 


int ceph_frag_compare(__u32 a, __u32 b)
{
        unsigned va = ceph_frag_value(a);
        unsigned vb = ceph_frag_value(b);
        if (va < vb)
                return -1;
        if (va > vb)
                return 1;
        va = ceph_frag_bits(a);
        vb = ceph_frag_bits(b);
        if (va < vb)
                return -1;
        if (va > vb)
                return 1;
        return 0;
}

Reply:
updated